### PR TITLE
[GH-99] T5.1.1 added configuration to logging feature

### PR DESCRIPTION
**Pull request by CodiePP**
added access to logging configuration in `LoggingFeature`

needed by https://github.com/input-output-hk/iohk-monitoring-framework/issues/228

issue #99

basis for `cardano-ledger` PR: https://github.com/input-output-hk/cardano-ledger/pull/487

### CODE REVIEW

File: app/Cardano/Shell/Features/Logging.hs [hunk 1]
```diff
@@ -94,13 +95,14 @@ loggingCardanoFeatureInit loggingConfig = do
...
-                    , llLogError    = Trace.logError
-                    , llAppendName  = Trace.appendName
+                    { llBasicTrace    = Trace.natTrace liftIO baseTrace
+                    , llLogDebug      = Trace.logDebug
+                    , llLogInfo       = Trace.logInfo
+                    , llLogNotice     = Trace.logNotice
+                    , llLogWarning    = Trace.logWarning
+                    , llLogError      = Trace.logError
+                    , llAppendName    = Trace.appendName
+                    , llConfiguration = lpConfiguration loggingConfig
```
Comment (ksaric) [inline app/Cardano/Shell/Features/Logging.hs]: This looks kind of weird. 
Isn't it possible to add the `bracket` function here and export it?

Comment (CodiePP) [reply app/Cardano/Shell/Features/Logging.hs]: yes, I am experimenting with exporting everything that is needed in the target app and get rid of the direct import of the logging library. Types and functions should be (re-)exported through the feature in `node-shell`.

File: app/Cardano/Shell/Features/Logging.hs [hunk 2]
```diff
@@ -48,13 +56,16 @@ data LoggingParameters = LoggingParameters
...
-    , llLogError   :: forall m a. (MonadIO m, Show a) => Trace m a  -> a -> m ()
-    , llAppendName :: forall m a. (MonadIO m, Show a) => LoggerName -> Trace m a -> m (Trace m a)
+    { llBasicTrace      :: forall m . MonadIO m => Trace m Text
+    , llLogDebug        :: forall m a. (MonadIO m, Show a) => Trace m a  -> a -> m ()
+    , llLogInfo         :: forall m a. (MonadIO m, Show a) => Trace m a  -> a -> m ()
+    , llLogNotice       :: forall m a. (MonadIO m, Show a) => Trace m a  -> a -> m ()
+    , llLogWarning      :: forall m a. (MonadIO m, Show a) => Trace m a  -> a -> m ()
+    , llLogError        :: forall m a. (MonadIO m, Show a) => Trace m a  -> a -> m ()
+    , llAppendName      :: forall m a. (MonadIO m, Show a) => LoggerName -> Trace m a -> m (Trace m a)
+    , llConfiguration   :: Configuration
```
Comment (ksaric) [inline app/Cardano/Shell/Features/Logging.hs]: Like I said, I don't think we need a `Configuration`.

File: app/Cardano/Shell/Features/Logging.hs [hunk 3]
```diff
@@ -48,13 +56,16 @@ data LoggingParameters = LoggingParameters
...
+    , llBracketMonadX   :: forall m a t. (MonadIO m, Show a) => Configuration -> Trace m a -> Severity -> Text -> m t -> m t
```
Comment (ksaric) [inline app/Cardano/Shell/Features/Logging.hs]: Formatting is messed up.

File: app/Cardano/Shell/Features/Logging.hs [hunk 4]
```diff
@@ -48,13 +56,16 @@ data LoggingParameters = LoggingParameters
...
+    , llBracketStmIO    :: forall a t. (Show a) => Configuration -> Trace IO a -> Severity -> Text -> STM.STM t -> IO t
```
Comment (ksaric) [inline app/Cardano/Shell/Features/Logging.hs]: Formatting.

File: cardano-shell.cabal [hunk 5]
```diff
@@ -45,30 +45,28 @@ library
     , specs
   build-depends:
       aeson
-    , base >=4.7 && <5
+    , base >=4.11 && <5
```
Comment (ksaric) [inline cardano-shell.cabal]: Why is this changed?

File: cardano-shell.cabal [hunk 6]
```diff
@@ -45,30 +45,28 @@ library
...
     , binary
     , bytestring
-    , cardano-prelude
     , Cabal
+    , cardano-prelude
```
Comment (ksaric) [inline cardano-shell.cabal]: I don't like the fact that people are reordering dependencies and removing comments from them.

File: cardano-shell.cabal [hunk 7]
```diff
@@ -45,30 +45,28 @@ library
...
-    , dhall
-    , ekg-core
     , process
     , QuickCheck
+    , safe-exceptions
+    , stm
     , text
     , transformers
-    , unix
+  if os(windows)
```
Comment (ksaric) [inline cardano-shell.cabal]: Good point, the path we were going with was "let's have things working on *nix and we'll deal with Windows later".

File: src/Cardano/Shell/Types.hs [hunk 8]
```diff
@@ -17,7 +17,7 @@ import           Control.Concurrent.Classy (MonadConc)
 
 import           Cardano.Shell.Constants.Types (CardanoConfiguration (..))
 
-import qualified System.Metrics as Ekg
+-- import qualified System.Metrics as Ekg
```
Comment (ksaric) [inline src/Cardano/Shell/Types.hs]: Do we need these comments? EKG is not really an issue, the stack traces just indicate something went wrong, not that it's crashing the whole application.

Comment (erikd) [reply src/Cardano/Shell/Types.hs]: Commented out code should never be merged to master.

File: app/Cardano/Shell/Features/Logging.hs (hunk 2 above)
Comment (erikd) [reply app/Cardano/Shell/Features/Logging.hs]: The code in `cardano-ledger` currently uses the `llConfiguration` field.

File: cardano-shell.cabal (hunk 5 above)
Comment (erikd) [reply cardano-shell.cabal]: Most of the `cardano-*` projects only build with ghc-8.4 and later. Ghc-8.4 ships with `base-4.11.*`

Comment (CodiePP) [reply cardano-shell.cabal]: I tried to get rid of the contravariant package. It is included in base-4.12.
But, couldn't raise the version too high. Can be reverted, unless prelude is updated.

File: cardano-shell.cabal (hunk 7 above)
Comment (CodiePP) [reply cardano-shell.cabal]: this is required by the iohk-nix cross-compilation setup in cardano-ledger.
